glide-core/redis-rs/redis/src/commands/cluster_scan.rs

```diff
@@ -284,467 +478,406 @@ impl ScanState {
         scan_state = scan_state
-            .create_updated_scan_state_for_completed_address(&core)
+            .create_updated_scan_state_for_completed_address(core, allow_non_covered_slots)
             .await?;
     }
 
-    // If the address is empty, meaning we finished scanning all the address
+    // Verify if the entire cluster has been scanned
     if scan_state.scan_status == ScanStateStage::Finished {
+        // Return the final scan state and the collected keys
         return Ok((ScanStateRC::create_finished(), new_keys));
     }
 
+    // Update the scan state with the new cursor and maintain the progress
     scan_state = ScanState::new(
         new_cursor,
         scan_state.scanned_slots_map,
         scan_state.address_in_scan,
         scan_state.address_epoch,
         ScanStateStage::InProgress,
     );
+
+    // Return the updated scan state and the newly found keys
     Ok((ScanStateRC::from_scan_state(scan_state), new_keys))
 }
 
-// Send the scan command to the address in the scan_state
+/// Sends the `SCAN` command to the specified address.
+///
+/// # Arguments
+///
+/// * `scan_state` - The current scan state.
+/// * `cluster_scan_args` - Arguments for the scan operation, including match pattern, count, object type, and allow_non_covered_slots.
+/// * `core` - The cluster connection.
+///
+/// # Returns
+///
+/// A `RedisResult` containing the response from the `SCAN` command.
 async fn send_scan<C>(
     scan_state: &ScanState,
-    core: &C,
-    match_pattern: Option<Vec<u8>>,
-    count: Option<usize>,
-    object_type: Option<ObjectType>,
+    cluster_scan_args: &ClusterScanArgs,
+    core: Arc<InnerCore<C>>,
 ) -> RedisResult<Value>
 where
-    C: ClusterInScan,
+    C: ConnectionLike + Connect + Clone + Send + Sync + 'static,
 {
-    let mut scan_command = cmd("SCAN");
-    scan_command.arg(scan_state.cursor);
-    if let Some(match_pattern) = match_pattern {
-        scan_command.arg("MATCH").arg(match_pattern);
-    }
-    if let Some(count) = count {
-        scan_command.arg("COUNT").arg(count);
-    }
-    if let Some(object_type) = object_type {
-        scan_command.arg("TYPE").arg(object_type.to_string());
+    if let Some(conn_future) = core
+        .connection_for_address(&scan_state.address_in_scan)
+        .await
+    {
+        let mut conn = conn_future.await;
+        let mut scan_command = cmd("SCAN");
+        scan_command.arg(scan_state.cursor);
+        if let Some(match_pattern) = cluster_scan_args.match_pattern.as_ref() {
+            scan_command.arg("MATCH").arg(match_pattern);
+        }
+        if let Some(count) = cluster_scan_args.count {
+            scan_command.arg("COUNT").arg(count);
+        }
+        if let Some(object_type) = &cluster_scan_args.object_type {
+            scan_command.arg("TYPE").arg(object_type.to_string());
+        }
+        conn.req_packed_command(&scan_command).await
+    } else {
+        Err(RedisError::from((
```

Comment:
https://github.com/valkey-io/valkey-glide/pull/2815#discussion_r1888485775

Please add some more info to this error message.

format!("Cluster scan failed. No connection available for address {}", address_in_scan)